text/0143-sdk-fail-safe-mode.md

```diff
@@ -0,0 +1,524 @@
+
+## Idea 1: Out of process crash detection <a name="idea-1"></a>
+
+The SDK launches an extra process to monitor and detect a crash in the user’s application. The main advantage is that when the SDK running in the extra process crashes, it doesn’t impact the user’s application process. While this seems appealing, it’s not possible on iOS and Android when writing this, and therefore, we can discard this option.
+
+## Idea 2: Detecting Crashing While Writing a Crash Report <a name="idea-2"></a>
+
+We could detect crashing while writing a crash report by writing a minimal crash report with only bare-bone information and then a full crash report with more details. If the bare-bone crash report misses its counterpart, the complete crash report, we know something is broken.
+
+This has other benefits, too. For example, it would help improve the watchdog termination detection algorithm of the Cocoa SDK.
+
+## Idea 3: Educate Users <a name="idea-3"></a>
+
+Our users could have avoided the [incident mentioned above](#background) by using a staged rollout. While we can't expect all users to use a staged rollout, we could explain actions to increase stability in a section in our docs, helping our users to ship more confidently. We can explain how they can use Sentry in a staged rollout to identify other severe problems, including continuous crashes of any type.
+
+## Idea 4: Query Play Store or App Store Connect for SDK crashes <a name="idea-4"></a>
+
+Our backend could query the Google Play Console or the App Store Connect APIs for Sentry SDK crashes. We need to find a use case for our users to link those APIs to Sentry. The major downside of these crashes is that they lack all scope information, such as breadcrumbs, tags, user info, device info, etc.,  and we can't connect them to other Sentry data, such as traces, replays, etc. 
+The Google Play Console provides the required API, but the App Store Connect API doesn't. The closest Apple offers Power and Performance Metric Logs, but you can't get crash reports via that API. We might be able to use spaceship.get_beta_feedback, though.
+
+# FAQ
+
+**1. Why do we make the SDK init a NoOp instead of trying to detect which integration is causing the crash?**
+
+We want to keep the logic as simple as possible to reduce the risk of bugs, and it could be that we disable the wrong integration, which would cause more unnecessary crashes.
+
+**2. Why do we focus mostly on crashes during and after SDK init?**
+
+As described in the [motivation](#motivation), continuous SDK crashes during the app start are more critical because they cause the most friction by making the application unusable.
+
+**3. Do we allow users to opt-out of this feature?**
+
+Yes, we can add a flag to the SDK to opt-out of this feature.
+
+**4. Do we allow a configuration for number of retries for our users?**
+
+No, a configuration adds complexity and we must avoid it. We only allow to enable or disable the feature.
```

Comment:
Do we want to allow configuration by customers on whether Safe Mode should be used or whether the SDK should go into No-Op directly?

Reply:
No, I don't have that planned to keep it simple. It's all or nothing. Do you think we need that? If yes, what is the use case?

Reply:
I was just thinking about how many chances I'd give an app that crashes on startup and it might be nice to let customers choose their priority (i.e. disable Sentry on a startup crash). However we can revisit if customers actually ask for this.

Reply:
> However we can revisit if customers actually ask for this.

Yes indeed.